lib/spack/spack/build_systems/lua.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+
+import os
+
+from llnl.util.filesystem import find
+
+from spack.directives import depends_on, extends
+from spack.package import PackageBase, run_before
+from spack.util.executable import Executable
+
+
+class LuaPackage(PackageBase):
+    """Specialized class for lua packages"""
+
+    phases = ['unpack', 'preprocess', 'install']
+    #: This attribute is used in UI queries that need to know the build
+    #: system base class
+    build_system_class = 'LuaPackage'
+
+    list_depth = 1  # LuaRocks requires at least one level of spidering to find versions
+    depends_on('lua-lang')
+    extends('lua', when='^lua')
+    extends('lua-luajit', when='^lua-luajit')
+    depends_on('luajit', when='^lua-luajit')
+    depends_on('luajit', when='^lua-luajit-openresty')
+    depends_on('lua-luajit+lualinks', when='^lua-luajit')
+    extends('lua-luajit-openresty', when='^lua-luajit-openresty')
+    depends_on('lua-luajit-openresty+lualinks', when='^lua-luajit-openresty')
+
+    def unpack(self, spec, prefix):
+        if os.path.splitext(self.stage.archive_file)[1] == '.rock':
+            directory = self.luarocks('unpack', self.stage.archive_file, output=str)
+            dirlines = directory.split('\n')
+            os.chdir(dirlines[2])
```

Comment:
Can we store the directory and then use the context manager `working_dir` where appropriate? That should restore the current working directory on exit from the context.

Reply:
We can certainly store the directory, but I've had zero luck figuring out how to use `with working_dir` in an un-scoped manner like this needs.  Is it possible to set the `build_directory` dynamically in an early phase and tell spack to re-evaluate it or something? I agree that I'm unfond of this but hadn't come up with a better solve.

Reply:
If that is hard to do, I would block merging this PR. Can you leave a `TODO` comment though or something like that?

Reply:
Sure, adding one now.